libparsec/crates/client/src/workspace/merge.rs

```diff
@@ -186,31 +186,68 @@ pub(super) fn merge_local_folder_manifest(
     // 0) Sanity checks, caller is responsible to handle them properly !
     debug_assert_eq!(local_base_id, &remote.id);
 
-    // TODO: Allow to force re-applying the prevent sync pattern (idempotent)
-    // if force_apply_pattern {
-    //     local_manifest = local_manifest.apply_prevent_sync_pattern(prevent_sync_pattern, timestamp)
-    // }
-
     // 1) Shortcut in case the remote is outdated
     if remote.version <= *local_base_version {
         return MergeLocalFolderManifestOutcome::NoChange;
     }
 
-    let local_from_remote = LocalFolderManifest::from_remote_with_local_context(
-        remote,
-        prevent_sync_pattern,
-        local,
-        timestamp,
-    );
+    // TODO: do we need this step ?
+    // // 2) Allow to force re-applying the prevent sync pattern (idempotent)
+    // //
+    // // The prevent sync pattern can change at any time, which in turn triggers
+    // // a full refresh of all local manifest to update their confinement points.
+    // // Here we make sure the local manifest doesn't have its confinement points
+    // // outdated, which would cause inconsistency in the next step.
+    // let local = local.apply_prevent_sync_pattern(prevent_sync_pattern, timestamp);
+
+    // 3) Confinement points is a special case: given any entry that match the prevent
+    // sync pattern is kept appart, there is no possibility of conflict here (i.e.
+    // confined local entries stay in `local.children`, confined remote entries stay in
+    // `local.base.children`).
+    //
+    // So if the current local manifest only changes are confined entries, then the merge
+    // operation only consist of converting the remote into a local and adding those
+    // confined entries back.
+    // This is precisely what we are doing in this step, and why the resulting manifest
+    // is called `merge_in_progress`: if the local manifest contains other changes, then
+    // we must do a proper children merge and update the manifest accordingly.
+
+    let mut merge_in_progress =
+        LocalFolderManifest::from_remote_with_restored_local_confinement_points(
+            remote,
+            prevent_sync_pattern,
+            local,
+            timestamp,
+        );
+    let remote = &merge_in_progress.base;
```

Comment:
Given this code is about merging a local and a remote, I think it's clearer to refer as `remote` instead of `merge_in_progress.base` when doing things like:
```rust
    let merged_parent = merge_parent(*local_base_parent, *local_parent, remote.parent);
```

Reply:
I agree with `assert_eq!(remote, merge_in_progress.base)`, however this code is about merging `remote` and `local`.

So I think it's more readable to see:
```rust
    let merge_need_sync = merged_children != remote.children || merged_parent != remote.parent;
```
given it's simply obvious we are comparing the merged data with the remote data.

While doing things like `merged_children != merge_in_progress.base.children` is more convoluted given you have one more conversion to do in your head.